File: apps/desktop/src/lib/tauri/tag.ts
import type { CreateTagInput, Tag, UpdateTagInput } from "@/types/tag";

import { invokeCommand } from "./client";

export const tagApi = {
	list() {
		return invokeCommand<Tag[]>("tag_list");
	},

	create(input: CreateTagInput) {
		return invokeCommand<Tag>("tag_create", {
			input: {
				name: input.name,
				color: input.color,
			},
		});
	},

	delete(id: string) {
		return invokeCommand<void>("tag_delete", { id });
	},

	update(id: string, input: UpdateTagInput) {
		return invokeCommand<Tag>("tag_update", {
			id,
			input: {
				color: input.color,
			},
		});
	},
};

File: apps/desktop/src-tauri/src/commands/tag.rs
use tauri::State;

use crate::application::services::TagService;
use crate::domain::tag::{CreateTagInput, TagDto};
use crate::error::AppError;
use crate::state::AppState;

#[tauri::command]
pub async fn tag_list(state: State<'_, AppState>) -> Result<Vec<TagDto>, AppError> {
    TagService::list(&state).await
}

#[tauri::command]
pub async fn tag_create(
    state: State<'_, AppState>,
    input: CreateTagInput,
) -> Result<TagDto, AppError> {
    TagService::create(&state, input).await
}

#[tauri::command]
pub async fn tag_delete(state: State<'_, AppState>, id: String) -> Result<(), AppError> {
    TagService::delete(&state, &id).await
}
